tiny_dnn/activations/softmax_layer.h

```diff
@@ -0,0 +1,49 @@
+/*
+    Copyright (c) 2017, Taiga Nomi, Karan Desai
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+#include "tiny_dnn/activations/activation_layer.h"
+#include "tiny_dnn/layers/layer.h"
+
+namespace tiny_dnn {
+
+class softmax_layer : public activation_layer {
+ public:
+  using activation_layer::activation_layer;
+
+  std::string layer_type() const override { return "softmax-activation"; }
+
+  void forward_activation(const vec_t &x, vec_t &y) {
+    float_t alpha = *std::max_element(x.begin(), x.end());
+    float_t denominator(0);
+    for (serial_size_t j = 0; j < x.size(); j++) {
+      y[j] = std::exp(x[j] - alpha);
+      denominator += y[j];
+    }
+    for (serial_size_t j = 0; j < x.size(); j++) {
+      y[j] /= denominator;
+    }
+  }
+
+  void backward_activation(const vec_t &x,
+                           const vec_t &y,
+                           vec_t &dx,
+                           const vec_t &dy) {
+    const serial_size_t len = static_cast<serial_size_t>(dy.size());
+
+    // auxilliary vector to store element wise softmax gradients of all elements
+    vec_t df(len, 0);
```

Comment:
Adding `thread_local` specifier to `vec_t df` local variable and resizing it later might improve performance as it reduces memory allocation/deallocation. This method will frequently be called during training.